backend/flags/flags_test.go

```diff
@@ -0,0 +1,99 @@
+package flags
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/heetch/confita"
+	"github.com/stretchr/testify/require"
+)
+
+type Config struct {
+	A    string        `config:"a"`
+	Adef string        `config:"a-def"`
+	B    bool          `config:"b"`
+	Bdef bool          `config:"b-def"`
+	C    time.Duration `config:"c"`
+	Cdef time.Duration `config:"c-def"`
+	D    int           `config:"d"`
+	Ddef int           `config:"d-def"`
+}
+
+func runHelper(t *testing.T, args ...string) *Config {
+	t.Helper()
+
+	var output bytes.Buffer
+
+	cs := []string{"-test.run=TestHelperProcess", "--"}
+	cs = append(cs, args...)
+	cmd := exec.Command(os.Args[0], cs...)
+	cmd.Stderr = &output
+	cmd.Env = []string{"GO_HELPER_PROCESS=1"}
+	err := cmd.Run()
+	require.NoError(t, err)
+
+	var cfg Config
+
+	err = json.NewDecoder(&output).Decode(&cfg)
+	require.NoError(t, err)
+
+	return &cfg
+}
+
+func TestFlags(t *testing.T) {
+	t.Run("Use defaults", func(t *testing.T) {
+		cfg := runHelper(t, "-a=hello", "-b=true", "-c=10s", "-d=-100")
+		require.Equal(t, "hello", cfg.A)
+		require.Equal(t, true, cfg.B)
+		require.Equal(t, 10*time.Second, cfg.C)
+		require.Equal(t, -100, cfg.D)
+	})
+
+	t.Run("Override defaults", func(t *testing.T) {
+		cfg := runHelper(t, "-a-def=bye", "-b-def=false", "-c-def=15s", "-d-def=-200")
+		require.Equal(t, "bye", cfg.Adef)
+		require.Equal(t, false, cfg.Bdef)
+		require.Equal(t, 15*time.Second, cfg.Cdef)
+		require.Equal(t, -200, cfg.Ddef)
+	})
+}
+
+func TestHelperProcess(t *testing.T) {
+	if os.Getenv("GO_HELPER_PROCESS") != "1" {
+		return
+	}
+
+	args := os.Args
+	for len(args) > 0 {
+		if args[0] == "--" {
+			args = args[1:]
+			break
+		}
+		args = args[1:]
+	}
+	if len(args) == 0 {
+		fmt.Fprintf(os.Stderr, "No args\n")
+		os.Exit(2)
+	}
+
+	os.Args = append(os.Args[:1], args...)
```

Comment:
What is the purpose of this line ?

Reply:
`flag.Parse` parses `os.Args`. The binary used to run this function is used with several arguments but the way the arguments are passed is not what we want Confita to parse.

Here is the command produced by the test "Use Defaults":

```
/tmp/go-build500277604/b001/flags.test -test.run=TestHelperProcess -- -a=hello -b=true -c=10s -d=-100
```

You can notice that between the `-test.run` flag and the `-a -b` flags there is `--`

This is to indicate the test binary "don't parse anything after `--`".
Without that we would have this error:

```sh
$ /tmp/go-build500277604/b001/flags.test -test.run=TestHelperProcess -a=hello -b=true -c=10s -d=-100

flag provided but not defined: -a
Usage of /tmp/go-build605850438/b001/flags.test:
  -httptest.serve string
    	if non-empty, httptest.NewServer serves on this address and blocks
  -test.bench regexp
    	run only benchmarks matching regexp
...
```

With the `--` flags we are basically *serializing* the parameters

Once the binary is called, the `TestHelperProcess` is called and its os.Args slice contains this:

`[/tmp/go-build189961190/b001/flags.test -test.run=TestHelperProcess -- -a=hello -b=true -c=10s -d=-100]`

`flag.Parse` is used by our backend to parse `os.Args`, but the content above doesn't represent a real world example like this one 

```sh
myprog -a=hello -b=true -c=10s -d=-100
```

We need to transform the content of os.Args to do that, that's what is done by the loop and the append. After that we have this content in os.Args:

```
[/tmp/go-build611018318/b001/flags.test -a-def=bye -b-def=false -c-def=15s -d-def=-200]
```

We can then call the backend and ask it to parse.

Reply:
Thanks for this deep explanation.